### PR TITLE
fix: compatible with react 17

### LINKED ISSUE #46256: xport 'useId' (imported as 'useId') was not found in 'react' With react 17
### Reproduction link

[Edit on CodeSandbox](https://codesandbox.io/p/sandbox/antd-reproduction-template-forked-cwz75t?file=%2Fpackage.json%3A12%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpr88iaq0006356hgv1two7v%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpr88iap0002356hqqs7hp3g%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpr88iap0003356hysup2q9p%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpr88iap0005356hghmqol5v%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpr88iap0002356hqqs7hp3g%2522%253A%257B%2522id%2522%253A%2522clpr88iap0002356hqqs7hp3g%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpr89kz2004m356hwt0eql8a%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A12%252C%2522startColumn%2522%253A1%252C%2522endLineNumber%2522%253A12%252C%2522endColumn%2522%253A1%257D%255D%252C%2522filepath%2522%253A%2522%252Fpackage.json%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clpr89kz2004m356hwt0eql8a%2522%257D%252C%2522clpr88iap0005356hghmqol5v%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpr88iap0004356hzdr4pu0n%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%2522%257D%255D%252C%2522id%2522%253A%2522clpr88iap0005356hghmqol5v%2522%252C%2522activeTabId%2522%253A%2522clpr88iap0004356hzdr4pu0n%2522%257D%252C%2522clpr88iap0003356hysup2q9p%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpr88iap0003356hysup2q9p%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

### Steps to reproduce

Latest ant-d with react 17.0  fails to compile

### What is expected?

compile error should not occur

### What is actually happening?

ERROR in ./node_modules/antd/es/config-provider/hooks/useThemeKey.js 3:27-32
export 'useId' (imported as 'useId') was not found in 'react' (possible exports: Children, Component, Fragment, Profiler, PureComponent, StrictMode, Suspense, __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, cloneElement, createContext, createElement, createFactory, createRef, forwardRef, isValidElement, lazy, memo, useCallback, useContext, useDebugValue, useEffect, useImperativeHandle, useLayoutEffect, useMemo, useReducer, useRef, useState, version)
 
ERROR in ./node_modules/antd/es/config-provider/hooks/useThemeKey.js 3:64-69
export 'useId' (imported as 'useId') was not found in 'react' (possible exports: Children, Component, Fragment, Profiler, PureComponent, StrictMode, Suspense, __SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED, cloneElement, createContext, createElement, createFactory, createRef, forwardRef, isValidElement, lazy, memo, useCallback, useContext, useDebugValue, useEffect, useImperativeHandle, useLayoutEffect, useMemo, useReducer, useRef, useState, version)
 
webpack compiled with 2 errors and 5 warnings

| Environment | Info |
| --- | --- |
| antd | 5.12.1 |
| React | react 17 |
| System | Windows 11 |
| Browser | Chrome |

### PR DESCRIPTION
**Pull request by MadCcc**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link
close https://github.com/ant-design/ant-design/issues/46256

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |   MISC: Fix `useId` error when webpack build with React 17        |
| 🇨🇳 Chinese |   MISC: 修复 React 17 以下使用 webpack 构建时报错 `useId` 找不到的问题。        |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

### CODE REVIEW

**Review by yoyo837 — APPROVED** [review]

File: components/config-provider/hooks/useThemeKey.ts [hunk 1]
```diff
@@ -1,4 +1,9 @@
-import { useId } from 'react';
+import * as React from 'react';
+
+const fullClone = {
+  ...React,
+};
+const { useId } = fullClone;
```
Comment (yoyo837) [inline components/config-provider/hooks/useThemeKey.ts:6]: 这里能 `React. useId` 吗？

Comment (afc163) [reply components/config-provider/hooks/useThemeKey.ts:6]: 为啥 ci 检查不出来？

**Review by li-jia-nan — APPROVED** [review]

File: components/config-provider/hooks/useThemeKey.ts (hunk 1 above)
Comment (MadCcc) [reply components/config-provider/hooks/useThemeKey.ts:6]: > 为啥 ci 检查不出来？

起了个 umi 项目试了下其实不会报错，不知道是不是 webpack 配置问题

**Comment by yoyo837** [comment]
啊， 这个menu 用例咋通过的

**Comment by guoshencheng** [comment]
when will it be released?

**Comment by fz6m** [comment]
如果某个方法没有被导出，webpack 会发出没有 export 的警告，umi 之所以不提示是因为把 warning 都禁了。

可以尝试如下方法越过 webpack 检查：

```ts
import * as ReactExports from 'react'

const IDENT = 'useId'
const useId = IDENT in ReactExports ? ReactExports[IDENT] : fallback
```

**Pull request merged**